connectome_interpreter/utils.py

```diff
@@ -1087,3 +1088,180 @@ def path_for_ngl(path):
+        # Extract the selected index values
+        values = grid_outputs[:, selected_index,
+                              i].mean(axis=1)
+
+        # Reshape values to match the grid
+        grid_size = int(np.sqrt(len(grid_coords)))
+        values_grid = values.reshape((grid_size, grid_size))
+        heatmaps.append(values_grid)
+
+    slider = widgets.IntSlider(value=1, min=1, max=len(heatmaps), step=1,
+                               description='Timepoint', continuous_update=True)
+
+    # Link the slider to the plotting function
+    display(widgets.interactive(plot_heatmap, index=slider,
+            xlab=xlab, ylab=ylab, title=title))
+
+
+@dataclass
+class XORCircuit:
+    # Input nodes
+    input1: list
+    input2: list
+    # Middle layer nodes
+    exciter1: str  # Takes input from input1 only
+    exciter2: str  # Takes input from input2 only
+    inhibitor: str        # Takes input from both inputs
+    # Output node
+    output: list
+
+
+def find_xor(paths: pd.DataFrame) -> List[XORCircuit]:
+    """
+    Find XOR-like circuits in a 3-layer network, based on [Wang et al. 2024](https://www.biorxiv.org/content/10.1101/2024.09.24.614724v2). Note: this function currently ignores middle excitatory neruons that receive both inputs.
+
+    Args:
+        paths: DataFrame with columns ['pre', 'post', 'sign', 'layer']
+        pre: source node
+        post: target node
+        sign: 1 (excitatory) or -1 (inhibitory)
+        layer: 1 (input->middle) or 2 (middle->output)
+
+    Returns:
+    List of XORCircuit objects, each representing a found XOR motif
+    """
+    # checking input ----
+    if set(paths.layer) != {1, 2}:
+        raise ValueError(
+            "The input DataFrame should have exactly 2 unique layers, 1 and 2")
+
+    # check column names of paths
+    if not {'pre', 'post', 'sign', 'layer'}.issubset(paths.columns):
+        raise ValueError(
+            "The input DataFrame should have columns ['pre', 'post', 'sign', 'layer']")
+
+    # check values of signs: if it's a subset of {1, -1}
+    if set(paths.sign) > {1, -1}:
+        raise ValueError(
+            f"The input DataFrame should have values 1 (excitatory) or -1 (inhibitory) in the 'sign' column, but got {set(paths.sign)}")
+
+    # define variables ----
+    circuits = []
+
+    exciters = paths.pre[(paths.layer == 2) & (paths.sign == 1)].unique()
+    inhibitors = paths.pre[(paths.layer == 2) & (paths.sign == -1)].unique()
+
+    l1 = paths[paths.layer == 1]
+    l2 = paths[paths.layer == 2]
+
+    exciter_us_dict = {exc: set(l1.pre[l1.post == exc]) for exc in exciters}
+    inhibitor_us_dict = {inh: set(l1.pre[l1.post == inh])
+                         for inh in inhibitors}
+
+    exciter_ds_dict = {exc: set(l2.post[l2.pre == exc]) for exc in exciters}
+    inhibitor_ds_dict = {inh: set(l2.post[l2.pre == inh])
+                         for inh in inhibitors}
+
+    # main algorithm ----
+    for e1, e2 in itertools.combinations(exciters, 2):
```

Comment:
You never need them the other way round? Or do you want `permutations` instead?

Reply:
Thanks! I don't want them the other way around actually. The two exciters are essentially symmetric in the function, and in what they are meant to do.